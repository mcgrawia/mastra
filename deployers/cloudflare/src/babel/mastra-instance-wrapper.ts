import type { PluginObj } from '@babel/core';
import * as babel from '@babel/core';

/**
 * Babel plugin that transforms Mastra exports for Cloudflare Workers compatibility.
 *
 * This plugin:
 * 1. Identifies named exports of the 'mastra' variable
 * 2. Checks if the export is a new instance of the 'Mastra' class
 * 3. Wraps the Mastra instantiation in an arrow function to ensure proper initialization
 *    in the Cloudflare Workers environment
 *
 * The transformation ensures the Mastra instance is properly scoped and initialized
 * for each request in the Cloudflare Workers environment.
 *
 * @returns {PluginObj} A Babel plugin object with a visitor that performs the transformation
 *
 * @example
 * // Before transformation:
 * export const mastra = new Mastra();
 *
 * // After transformation:
 * export const mastra = () => new Mastra();
 */
export function mastraInstanceWrapper(): PluginObj {
  const exportName = 'mastra';
  const className = 'Mastra';
  const t = babel.types;

  return {
    name: 'wrap-mastra',
    visitor: {
      ExportNamedDeclaration(path) {
        if (t.isVariableDeclaration(path.node?.declaration)) {
          for (const declaration of path.node.declaration.declarations) {
            if (
              t.isIdentifier(declaration?.id, { name: exportName }) &&
              t.isNewExpression(declaration?.init) &&
              t.isIdentifier(declaration.init.callee, { name: className })
            ) {
              declaration.init = t.arrowFunctionExpression([], declaration.init);
              // there should be only one "mastra" export, so we can exit the loop
              break;
            }
          }
        }
      },
    },
  } as PluginObj;
}
