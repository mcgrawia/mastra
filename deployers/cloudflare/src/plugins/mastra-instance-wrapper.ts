import { transformSync } from '@babel/core';
import type { Plugin } from 'rollup';
import { mastraInstanceWrapper as mastraInstanceWrapperBabel } from '../babel/mastra-instance-wrapper';

export function mastraInstanceWrapper(mastraEntryFile: string): Plugin {
  return {
    name: 'mastra-wrapper',
    transform(code, id) {
      if (id !== mastraEntryFile) {
        return null;
      }

      const result = transformSync(code, {
        filename: id,
        babelrc: false,
        configFile: false,
        plugins: [mastraInstanceWrapperBabel],
      });

      if (!result?.code) {
        throw new Error('mastra-wrapper plugin did not return code, there is likely a bug in the plugin.');
      }

      return {
        code: result.code,
        map: result?.map,
      };
    },
  };
}
