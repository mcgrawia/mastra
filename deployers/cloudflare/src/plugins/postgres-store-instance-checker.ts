import { transformSync } from '@babel/core';
import type { Plugin } from 'rollup';
import { postgresStoreInstanceChecker as postgresStoreInstanceCheckerBabel } from '../babel/postgres-store-instance-checker';

export function postgresStoreInstanceChecker(): Plugin {
  return {
    name: 'postgres-store-instance-checker',
    transform(code, id) {
      const result = transformSync(code, {
        filename: id,
        babelrc: false,
        configFile: false,
        plugins: [postgresStoreInstanceCheckerBabel],
      });

      if (!result?.code) {
        throw new Error(
          'postgres-store-instance-checker plugin did not return code, there is likely a bug in the plugin.',
        );
      }

      return {
        code: result.code,
        map: result?.map,
      };
    },
  };
}
