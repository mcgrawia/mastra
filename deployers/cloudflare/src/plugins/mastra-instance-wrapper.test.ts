import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { rollup } from 'rollup';
import esbuild from 'rollup-plugin-esbuild';
import { describe, it, expect } from 'vitest';
import { mastraInstanceWrapper } from '../plugins/mastra-instance-wrapper';

describe('Wrap Mastra instance', () => {
  const _dirname = dirname(fileURLToPath(import.meta.url));

  it.for([['./__fixtures__/empty-mastra.js'], ['./__fixtures__/basic.js']])(
    'should wrap Mastra instance in %s',
    async ([fileName]) => {
      const file = join(_dirname, fileName as string);

      const bundle = await rollup({
        logLevel: 'silent',
        input: file,
        cache: false,
        treeshake: 'smallest',
        plugins: [
          {
            name: 'externalize-all',
            resolveId(id) {
              return {
                id,
                external: id !== file,
              };
            },
          },
          esbuild({
            target: `esnext`,
            platform: 'browser',
            minify: false,
          }),
          mastraInstanceWrapper(file),
        ],
      });

      const result = await bundle.generate({
        format: 'esm',
      });

      expect(result?.output[0].code).toMatchSnapshot();
    },
  );
});
