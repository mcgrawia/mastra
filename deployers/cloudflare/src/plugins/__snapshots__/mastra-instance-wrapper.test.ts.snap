// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Wrap Mastra instance > should wrap Mastra instance in ./__fixtures__/basic.js 1`] = `
"import { Mastra } from '@mastra/core/mastra';
import { createLogger } from '@mastra/core/logger';
import { weatherAgent } from '@/agents';
import { TestDeployer } from '@mastra/deployer/test';
import { createApiRoute } from '@mastra/core/server';
import { PostgresStore } from '@mastra/pg';

const mastra = () => new Mastra({
  agents: {
    weatherAgent
  },
  storage: new PostgresStore({
    connectionString: process.env.DATABASE_URL
  }),
  logger: createLogger({
    name: "Mastra",
    level: "info"
  }),
  telemetry: {
    enabled: true,
    serviceName: "my-app",
    export: {
      type: "otlp",
      endpoint: "http://localhost:4318"
      // SigNoz local endpoint
    }
  },
  server: {
    port: 3e3,
    timeout: 5e3,
    apiRoutes: [createApiRoute({
      path: "/hello",
      method: "get",
      handler: async (req, res) => {
        res.send("Hello World");
      }
    })]
  },
  deployer: new TestDeployer()
});

export { mastra };
"
`;

exports[`Wrap Mastra instance > should wrap Mastra instance in ./__fixtures__/empty-mastra.js 1`] = `
"import { Mastra } from '@mastra/core/mastra';

const mastra = () => new Mastra();

export { mastra };
"
`;
